crates/uv/src/commands/project/run.rs

```diff
@@ -51,78 +51,118 @@ pub(crate) async fn run(
-        let lock = project::lock::do_lock(
-            project.workspace(),
-            venv.interpreter(),
-            settings.as_ref().into(),
-            preview,
-            connectivity,
-            concurrency,
-            native_tls,
-            cache,
-            printer,
-        )
-        .await?;
-        project::sync::do_sync(
-            project.project_name(),
-            project.workspace().root(),
-            &venv,
-            &lock,
-            extras,
-            dev,
-            Modifications::Sufficient,
-            settings.as_ref().into(),
-            preview,
-            connectivity,
-            concurrency,
-            native_tls,
-            cache,
-            printer,
-        )
-        .await?;
+        let venv = if let Some(project) = project {
+            debug!(
+                "Discovered project `{}` at: {}",
+                project.project_name(),
+                project.workspace().root().display()
+            );
+
+            let venv = project::init_environment(
+                project.workspace(),
+                python.as_deref().map(ToolchainRequest::parse),
+                toolchain_preference,
+                connectivity,
+                native_tls,
+                cache,
+                printer,
+            )
+            .await?;
+
+            // Lock and sync the environment.
+            let lock = project::lock::do_lock(
+                project.workspace(),
+                venv.interpreter(),
+                settings.as_ref().into(),
+                preview,
+                connectivity,
+                concurrency,
+                native_tls,
+                cache,
+                printer,
+            )
+            .await?;
+            project::sync::do_sync(
+                project.project_name(),
+                project.workspace().root(),
+                &venv,
+                &lock,
+                extras,
+                dev,
+                Modifications::Sufficient,
+                settings.as_ref().into(),
+                preview,
+                connectivity,
+                concurrency,
+                native_tls,
+                cache,
+                printer,
+            )
+            .await?;
+
+            venv
+        } else {
+            debug!("No project found; searching for Python interpreter");
+
+            let client_builder = BaseClientBuilder::new()
+                .connectivity(connectivity)
+                .native_tls(native_tls);
+
+            let toolchain = Toolchain::find_or_fetch(
+                python.as_deref().map(ToolchainRequest::parse),
+                EnvironmentPreference::Any,
+                toolchain_preference,
+                client_builder,
+                cache,
+            )
+            .await?;
+
+            PythonEnvironment::from_toolchain(toolchain)
+        };
 
         Some(venv)
     };
 
+    if let Some(project_env) = &project_env {
```

Comment:
It might make sense to rename `project_env` to something like `base_env` now that it can come from the system?